Introduce baseline JIT compiler

qwe661234 opened the pull request:
The baseline JIT compiler design involves tracing an EBB when its usage frequency exceeds a predetermined threshold. We then utilize a code generator to convert the instruction sequence into C code. The C code is subsequently compiled using mir, and the resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance the execution speed of RISC-V instructions with minimal modification. This implementation requires three additional components: a code generator, mir compiler, and code cache, while maintaining the original design of the interpreter. Furthermore, this baseline JIT compiler serves as the foundational target for future improvements.

|Metric|  master| baseline JIT|Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 1352.843 (Iterations/Sec)|1740.881 (Iterations/Sec) |+28.6％|
| dhrystone  |1146 DMIPS |2368.33DMIPS |+106%|

jserv commented on src/cache.c:
```diff
@@ -16,27 +17,31 @@
- * strategy, while T2 follows the LFU strategy. Additionally, ARC maintains two
- * ghost lists, B1 and B2, which store replaced entries from the LRU and LFU
- * lists, respectively.
+/*
```
Minimize the necessary changes. Don't change this comment.

jserv reviewed (changes requested):
Resolve the issues raised by CodeQL in advance.

jserv commented on src/cache.c:
```diff
@@ -16,27 +17,31 @@
- * exceeds the THRESHOLD, the JIT compiler flow is triggered.
- */
-#define THRESHOLD 1000
+ * exceeds the THRESHOLD, the JIT compiler flow is triggered. */
```
Minimize the necessary changes.

jserv commented on src/codegen.c:
```diff
@@ -0,0 +1,1298 @@
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[rv_insn_qty];
+static char funcbuf[128] = {0};
+#define RVOP(inst, code)                                                  \
```
Instead of duplicating `RVOP` and instruction executions, try to provide JIT-aware `RVOP` macro which concatenates the string for code generation.

jserv reviewed (changes requested):
Check the `jit_template.inc` in branch `wip/jit` and eliminate the code duplication between `execute.c` and `codegen.c`.

jserv commented on Makefile:
```diff
@@ -78,6 +78,21 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 1
```
Default to `0` (disabled).

jserv commented on Makefile:
```diff
@@ -78,6 +78,21 @@ gdbstub-test: $(BIN)
+mir/GNUmakefile:
+	git submodule update --init $(dir $@)
+	
+mir/libmir.a: mir/GNUmakefile
```
For the consistency of source code layout, place MIR directory under `src`. That is, specify `src/mir` here.

jserv commented on src/riscv_private.h:
```diff
@@ -119,8 +126,12 @@ struct riscv_internal {
+#if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *cache;
```
To minimize the code changes, is it possible to stick to `block_map_t` rather than `struct cache`?

qwe661234 replied:
It is difficult to stick to `block_map_t`, because the structure of ARC cache is totally different from `block_map_t`.

jserv commented on .vscode/settings.json:
```diff
@@ -0,0 +1,14 @@
+{
```
Don't check in anything related to `.vscode`.

jserv commented on Makefile:
```diff
@@ -109,9 +124,15 @@ $(OUT)/%.o: src/%.c
 	$(VECHO) "  LD\t$@\n"
 	$(Q)$(CC) -o $@ $^ $(LDFLAGS)
+else
+$(BIN): src/mir/libmir.a $(OBJS)
```
You can simply append `src/mir/libmir.a` to `$(OBJS)`.

jserv commented on src/codegen.c:
```diff
@@ -0,0 +1,1238 @@
+    strcat(gencode, "  pc = rv->PC;\n");
+    UPDATE_PC(ir->imm);
+    if (ir->rd) {
+        sprintf(funcbuf, "  rv->X[%u] = pc + %u;\n", ir->rd, ir->insn_len);
```
Don't invoke `sprintf` in each implementation of RISC-V instructions. Instead, wrap them in proper macros or inline functions.

jserv commented on src/elfdef.h:
```diff
@@ -0,0 +1,78 @@
+#include <stdint.h>
```
Don't create new file `elfdef.h`. Instead, consolidate the existing `elf.[ch]`.

jserv commented:
The proposed changes aim to achieve the following objectives:
1. Elimination of code duplication between `emulate.c` and `codegen.c`.
2. Reuse of the RV32 interpreter implementation for the baseline JIT.

These refinements focus on improving code organization and minimizing duplication. By reusing the existing RV32 interpreter implementation, we can expedite the development of the JIT compiler while maintaining consistency between the interpreter and JIT modes of execution.

The new branch `wip/jit` lays the groundwork for the future development of a baseline JIT compiler. By decoupling the RV32 instructions, we create reusable implementation templates that can be leveraged for code generation in the JIT compiler. This decoupling enables straightforward and flexible code generation strategies, eliminating the need for code duplication between the interpreter and MIR-based JIT compiler.

Assuming the existence of a dedicated shell script capable of generating C-string operations for each `RVOP` entry in `src/codegen.c` from `src/rv32_template.c`, the following procedure can be followed:
1. Enable the JIT compiler in the top-level configurations.
2. Generate `src/codegen.c` from `src/rv32_template.c` using the aforementioned script.
3. The file `src/emulate.c` includes the generated `src/codegen.c` for the JIT execution path.
4. The runtime determines the execution path, whether through the interpreter or JIT compilation, based on heuristic algorithms or thresholds.

This pull request should be on top of branch `wip/jit`, and you will have to rebase it and rework.

jserv commented on src/cache.c:
```diff
@@ -286,6 +305,12 @@ cache_t *cache_create(int size_bits)
 #endif
     cache->capacity = cache_size;
+#if RV32_HAS(JIT)
+    cache->jitcode = (uint8_t *) mmap(NULL, CODE_CACHE_SIZE,
```
The casting from `void *` to `uint8_t *` is not necessary in C programming language.

jserv reviewed (changes requested):
Rebase latest `wip/jit` branch rather than `master` branch.

jserv reviewed (changes requested):
Rebase latest `wip/jit` branch rather than `master` branch.

jserv commented on Makefile:
```diff
@@ -78,6 +78,26 @@ gdbstub-test: $(BIN)
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += codegen.o compile.o
+src/rv32_jit_template.c:
+	$(Q)python ./tools/jit_template_grnerator.py 
```
Rename to `tools/gen-jit-template.py`

jserv commented on src/riscv.h:
```diff
@@ -7,6 +7,22 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#if RV32_HAS(EXT_F)
```
Don't do that. The file `riscv.h` is supposed to be a public header, and there should be no implementation details.

jserv commented on .gitmodules:
```diff
@@ -4,3 +4,6 @@
 	path = src/mini-gdbstub
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "mir"]
+	path = mir
```
Should be `src/mir`.

jserv commented on .gitmodules:
```diff
@@ -4,3 +4,6 @@
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "src/mir"]
+	path = src/mir
+	url = https://github.com/vnmakarov/mir.git
```
Trim the trailing `.git` since it is not necessary.

jserv commented on Makefile:
```diff
@@ -78,6 +78,30 @@ gdbstub-test: $(BIN)
+$(OUT)/compile.o: src/compile.c src/mir/libmir.a
+	$(VECHO) "  CC\t$@\n"
+	$(Q)$(CC) -o $@ $(CFLAGS) -c -MMD -MF $@.d $<
+
```
Remove this blank line.

jserv commented on Makefile:
```diff
@@ -78,6 +78,30 @@ gdbstub-test: $(BIN)
+CFLAGS += -I./src/mir
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += codegen.o compile.o
+src/rv32_jit_template.c:
```
Simply call it `jit_template.c` for short.

jserv commented on src/cache.c:
```diff
@@ -226,6 +244,7 @@ static inline void hlist_del_init(struct hlist_node *n)
          pos = hlist_entry_safe((pos)->member.next, type, member))
 #endif
 
+
```
Don't do this. Minimize the necessary changes.

jserv commented on src/codegen.c:
```diff
@@ -0,0 +1,385 @@
+/*
```
Can you move the handwritten part of `src/codegen.c` into `src/compile.c`? Therefore, `src/codegen.c` would be completely generated by tools, implying that we can ignore it in git tree.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
```
Add copyright notice.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
```
Header `<sys/mman.h>` is not needed here.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,129 @@
+    for (item = DLIST_HEAD(MIR_item_t, modules->items); item; \
+         item = DLIST_NEXT(MIR_item_t, item))
+
+void *import_resolver(const char *name)
```
Mark it as `static`.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,129 @@
+static riscv_jit_t *jit = NULL;
+static jit_item_t *jit_ptr = NULL;
+
+uint8_t *compile(riscv_t *rv)
```
Is it an internal function? If so, mark it as `static`.

jserv commented on src/compile.h:
```diff
@@ -0,0 +1,12 @@
+#include <stdint.h>
+
+#include "riscv.h"
+#include "riscv_private.h"
```
Is header `riscv_private.h` needed?

jserv commented on src/decode.h:
```diff
@@ -171,6 +171,7 @@ enum {
 #define _(inst, can_branch) rv_insn_##inst,
     RISCV_INSN_LIST
 #undef _
+        rv_insn_qty,
```
Rename to `N_RV_INSN` for consistency.

jserv commented on src/riscv.c:
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
 }
+
+#if defined(__APPLE__)
+inline int isinff(float x)
```
Mark it as `static inline`.

jserv commented on src/riscv.c:
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
+{
+    return __builtin_fabsf(x) == __builtin_inff();
+}
+inline int isnanf(float x)
```
Ditto. Mark it as `static inline`.

jserv commented on src/riscv.h:
```diff
@@ -7,6 +7,16 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#if RV32_HAS(EXT_F)
```
Move these into `riscv_private.h`. Always ensure the simplicity of public headers.

jserv commented on src/riscv.h:
```diff
@@ -167,6 +177,26 @@ bool rv_has_halted(riscv_t *rv);
 /* return the flag of outputting exit code */
 bool rv_enables_to_output_exit_code(riscv_t *rv);
 
+#if RV32_HAS(Zicsr)
```
Ditto. Don't touch public headers.

jserv commented on tools/gen-jit-template.py:
```diff
@@ -0,0 +1,58 @@
+import operator
```
Enforce [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/).

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,507 @@
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static inline uint32_t hash(uint32_t key)
```
Why did you select this hash function?

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,507 @@
+    return (key >> 1) & (SET_SIZE - 1);
+}
+
+void set_reset(set_t *set)
```
Does it make sense to expose set related symbols?

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,507 @@
+        trace_ebb(rv, gencode, ir->branch_taken, set);
+}
+
+#define PROLOGUE                                                          \
```
Is it possible to generate `PROLOGUE` from C source file?

jserv commented on tools/gen-jit-template.py:
```diff
@@ -0,0 +1,156 @@
+
+f.close()
+# generate jit template
+f = open("src/jit_template.c", 'w')
```
Output to stdout by default.

jserv commented on tools/gen-jit-template.py:
```diff
@@ -0,0 +1,156 @@
+    "fuse3",
+    "fuse4"]
+# check enabled extension in Makefile
+f = open('Makefile', 'r')
```
Don't do that. Instead, the build rules should generate something corresponded to the current configuration. Then, tools can take the generated configuration file(s).

jserv commented on src/cache.c:
```diff
@@ -541,3 +568,117 @@ void cache_free(cache_t *cache, void (*callback)(void *))
+
+static inline uint64_t align_to(uint64_t val, uint64_t align)
+{
+    if (!align)
```
Add `unlikely` here.

jserv commented on src/emulate.c:
```diff
@@ -615,7 +504,27 @@ void rv_step(riscv_t *rv, int32_t cycles)
+        }
+        if (block->hot) {
+            /* execute machine code */
+            /* clang-format off */
```
Why did you add clang-format tweak here?

qwe661234 replied:
After using `clang-format-12`, this line would become `(exec_block_func_t) code) (rv, block->ir)`. However, it fails the coding style check because the expected format is without a space `(exec_block_func_t) code)(rv, block->ir)`.

jserv replied:
You can simply rewrite as following:
```c
exec_block_func_t _code = (exec_block_func_t) code;
if (unlikely(_code(rv, block->ir)))
```

jserv commented on src/riscv_private.h:
```diff
@@ -137,3 +148,33 @@ static inline uint32_t sign_extend_b(const uint32_t x)
+#include <math.h>
+#include "softfloat.h"
+
+#if defined(__APPLE__)
```
This part is no longer needed.

qwe661234 replied:
Except for Apple, the libraries of other computer architectures has provided the `isinff` and `isnanf` functions

jserv replied:
The following patch works for my Apple M1 machine:
```diff
--- a/src/riscv.c
+++ b/src/riscv.c
@@ -168,17 +168,6 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
     rv->halt = false;
 }
 
-#if defined(__APPLE__)
-static inline int isinff(float x)
-{
-    return __builtin_fabsf(x) == __builtin_inff();
-}
-static inline int isnanf(float x)
-{
-    return x != x;
-}
-#endif
-
 /* get current time in microsecnds and update csr_time register */
 static inline void update_time(riscv_t *rv)
 {
@@ -153,11 +153,15 @@ static inline uint32_t sign_extend_b(const uint32_t x)
 #include <math.h>
 #include "softfloat.h"
 
-#if defined(__APPLE__)
-inline int isinff(float x);
-inline int isnanf(float x);
+static inline int isinff(float x)
+{
+    return __builtin_fabsf(x) == __builtin_inff();
+}
+static inline int isnanf(float x)
+{
+    return x != x;
+}
 #endif
-#endif /* RV32_HAS(EXT_F) */
 
 #if RV32_HAS(Zicsr)
 /* CSRRW (Atomic Read/Write CSR) instruction atomically swaps values in the
```

qwe661234 replied:
It only works for the Apple architecture. In an x86 machine, the error occurs.
```
src/riscv_private.h:156:19: error: static declaration of 'isinff' follows non-static declaration
static inline int isinff(float x)
                  ^
src/riscv_private.h:160:19: error: static declaration of 'isnanf' follows non-static declaration
static inline int isnanf(float x)
                  ^
```

jserv replied:
> It only works for the Apple architecture. In an x86 machine, the error occurs.

Consolidate the conditional build upon system configurations.

jserv commented on src/riscv.c:
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
     rv->halt = false;
 }
+
+#if defined(__APPLE__)
```
Remove the inline functions `isinff` and `isnanf`. The function definitions should appear in `src/riscv_private.h`.

jserv commented:
TODO: benchmarking on the following machine:
* eMag 8180 Arm server w/ GNU/Linux
* Apple M1 (or other Apple Silicon) w/ macOS

We have to ensure the benefits.

jserv reviewed (changes requested):
1. Update CI pipeline to validate JIT-enabled execution.
2. Make comprehensive performance evaluations for various configurations. e.g., x86-64/linux, arm64/linux, arm64/macOS, etc.
3. Figure out performance bottleneck via profiling
4. Move forward to tracing JIT: https://en.wikipedia.org/wiki/Tracing_just-in-time_compilation

jserv commented on Makefile:
```diff
@@ -159,11 +184,11 @@ endif
+	$(RM) $(BIN) $(OBJS) $(deps) $(CACHE_OUT) src/jit_template.c
 distclean: clean
-	-$(RM) $(DOOM_DATA) $(QUAKE_DATA)
+	-$(RM) $(DOOM_DATA) $(QUAKE_DATA) 
```
Don't add trailing spaces.

jserv commented on src/cache.c:
```diff
@@ -8,6 +8,7 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/mman.h>
```
Windows lacks of header `sys/mman.h`. Therefore, apply conditional build.

jserv commented on src/jit_template.c:
```diff
@@ -0,0 +1,511 @@
+RVOP(nop, { strcat(gencode, "/* no operation */\n"); })
```
Don't check in `src/jit_template.c`.

jserv commented on Makefile:
```diff
@@ -78,6 +79,27 @@ gdbstub-test: $(BIN)
+OBJS_EXT += compile.o
+
+src/jit_template.c:
+	$(Q)python ./tools/gen-jit-template.py $(EXT_LIST) > src/jit_template.c
```
Replace with
```shell
$(Q)python tools/gen-jit-template.py $(EXT_LIST) > $@
```

jserv commented on src/emulate.c:
```diff
@@ -34,6 +23,11 @@ extern struct target_ops gdbstub_ops;
+#if RV32_HAS(JIT)
+#include "cache.h"
+#include "compile.h"
+typedef bool (*exec_block_func_t)(riscv_t *rv, rv_insn_t *ir);
```
Don't typedef here.

jserv commented on src/emulate.c:
```diff
@@ -615,7 +504,26 @@ void rv_step(riscv_t *rv, int32_t cycles)
-        /* execute the block */
+#if RV32_HAS(JIT)
+        /* execute the block by JIT compiler */
+        exec_block_func_t code = NULL;
```
Move tyepdef `exec_block_func_t` here.

qwe661234 replied:
I typedef it to the top of function `rv_step`, because formatter format it to `typedef bool (*exec_block_func_t)(riscv_t * rv, rv_insn_t * ir);` if we tyepdef here.

jserv commented on Makefile:
```diff
@@ -13,6 +13,7 @@ CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFFF000
 
 OBJS_EXT :=
 
+EXT_LIST :=
```
To avoid potential confusion, let's rename it to `RVEXT_LIST` instead of `EXT_LIST`. The latter may mislead someone into thinking it represents a list of file extensions.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+    char *name;
+    void *func;
+} func_obj_t;
+
```
Remove this blank line.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static void set_reset(set_t *set)
```
Decalare as a `static inline` function.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
```
Simply rename as
```c
while (set->table[index][count]) {
```

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
```
Ditto.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[N_RV_INSN];
+static char funcbuf[128] = {0};
+#define APPEND_CODE(...)           \
```
Rename it to `GEN`, which means generating code.

jserv commented on tools/gen-jit-template.py:
```diff
@@ -0,0 +1,209 @@
+import re
```
Add a general description for this script.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+#include "jit_template.c"
+
+RVOP(jal, {
+    strcat(gencode, "  pc = rv->PC;\n");
```
Provide a macro to encapsulate, so that `gencode` won't appear here.

jserv commented on Makefile:
```diff
@@ -78,6 +79,27 @@ gdbstub-test: $(BIN)
+OBJS_EXT += compile.o
+
+src/jit_template.c:
+	$(Q)python ./tools/gen-jit-template.py $(EXT_LIST) > $@
```
Rewrite as
```shell
$(Q)tools/gen-jit-template.py $(EXT_LIST) > $@
```

You shall specify a proper [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)) accordingly. i.e.,
```
#!/usr/bin/env python3
```

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+})
+#undef RVOP
+
+void trace_ebb(riscv_t *rv, char *gencode, rv_insn_t *ir, set_t *set)
```
Declare it as `static`.

jserv commented on src/cache.c:
```diff
@@ -20,7 +25,13 @@
  * exceeds the THRESHOLD, the JIT compiler flow is triggered.
  */
-#define THRESHOLD 1000
+#define THRESHOLD 10000
```
The change of `THRESHOLD` should come with an engineering consideration.

qwe661234 replied:
Perhaps we keep the value at 1000 for now and make modifications when we optimize the JIT.

jserv commented on src/riscv.c:
```diff
@@ -116,11 +123,23 @@ bool rv_enables_to_output_exit_code(riscv_t *rv)
 }
 
+#if RV32_HAS(JIT)
+void release_block(void *block)
```
Declare it as a `static` function.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,403 @@
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define GOLDEN_RATIO_32 0x61C88647
```
Both `GOLDEN_RATIO_32` and `HASH` already exist in `src/cache.c`. We shall put them to somewhere such as `src/utils.h`.

jserv commented:
It fails to build on macOS:
```shell
$ make  
Makefile:19: *** missing separator.  Stop.
```

System information:
```
$ make --version | head -1
GNU Make 3.81
```

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,402 @@
+
+#if RV32_HAS(EXT_C)
+RVOP(clw, {
+    GEN("    addr = rv->X[%u] + %u;\n", ir->rs1, ir->imm);
```
Be aware of the indention for C2MIR string.

jserv commented on src/riscv_private.h:
```diff
@@ -59,8 +61,12 @@ typedef struct block {
     struct block *predict;     /**< block prediction */
     rv_insn_t *ir;             /**< IR as memory blocks */
+#if RV32_HAS(JIT)
+    bool hot;
```
Add corresponding comments here.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,402 @@
+
+#include "jit_template.c"
+
+RVOP(jal, {
```
You have to explain why we have to provide the customized handlers for branch related instructions.

qwe661234 committed:
Introduce baseline JIT compiler

The baseline JIT compiler design involves tracing an EBB when its usage
frequency exceeds a predetermined threshold. We then utilize a code generator
to convert the instruction sequence into C code. The C code is subsequently
compiled using mir, and the resulting target machine code is stored in the
code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance
the execution speed of RISC-V instructions with minimal modification. This
implementation requires three additional components: a code generator, mir
compiler, and code cache, while maintaining the original design of the interpreter.
Furthermore, this baseline JIT compiler serves as the foundational target
for future improvements.

|  Metric   |        master           |       baseline JIT       |Speedup |
|-----------+-------------------------+--------------------------+--------|
| CoreMark  |1352.843 (Iterations/Sec)|1740.881 (Iterations/Sec) | +28.6% |
| dhrystone |1146 DMIPS               |2368.33 DMIPS             | +106%  |

qwe661234 commented:
> It fails to build on macOS:
> 
> ```shell
> $ make  
> Makefile:19: *** missing separator.  Stop.
> ```
> 
> System information:
> 
> ```
> $ make --version | head -1
> GNU Make 3.81
> ```

Solved

jserv commented:
Runtime failure on macOS/Arm64. Reproduce:
```shell
$ git clone https://github.com/sysprog21/rv32emu
$ cd rv32emu
$ gh pr checkout 135
$ make ENABLE_JIT=1
$ make check
```

Error messages:
```
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Running hello.elf ... Failed.
```

jserv commented:
This preliminary work was integrated into `wip/jit` branch for sake of synchronization with latest `master` branch.
@qwe661234, You shall work on `wip/jit` branch for all JIT specific issues and submit pull requests on top of it.
